plotter/sankey.go

```diff
@@ -0,0 +1,440 @@
+	// on the value axis.
+	min float64
+
+	// max is min plus the larger of receptorValue and sourceValue.
+	max float64
+
+	// sourceFlowPlaceholder and receptorFlowPlaceholder track
+	//  the current plotting location during
+	// the plotting process.
+	sourceFlowPlaceholder, receptorFlowPlaceholder float64
+}
+
+// A Flow represents the amount of an entity flowing between two stocks.
+type Flow struct {
+	// SourceStockLabel and ReceptorStockLabel are the labels
+	// of the stocks that originate and receive the flow,
+	// respectively.
+	SourceStockLabel, ReceptorStockLabel string
+
+	// SourceStockCategory and ReceptorStockCategory define
+	// the locations on the category axis of the stocks that
+	// originate and receive the flow, respectively. The
+	// SourceStockCategory must be a lower number than
+	// the ReceptorStockCategory.
+	SourceStockCategory, ReceptorStockCategory int
+
+	// Value represents the magnitute of the flow.
+	// It must be greater than or equal to zero.
+	Value float64
+
+	// Group specifies the group that a flow belongs
+	// to. It is used in assigning styles to groups
+	// and creating legends.
+	Group string
+}
+
+// NewSankey creates a new Sankey diagram with the specified
+// flows and stocks.
+func NewSankey(flows ...Flow) (*Sankey, error) {
+	s := new(Sankey)
+
+	s.stocks = make(map[int]map[string]*stock)
+
+	s.flows = flows
+	for i, f := range flows {
+		// check stock category order
+		if f.SourceStockCategory >= f.ReceptorStockCategory {
+			return nil, fmt.Errorf("plotter.NewSankey: Flow %d SourceStockCategory (%d) "+
+				">= ReceptorStockCategory (%d)", i, f.SourceStockCategory, f.ReceptorStockCategory)
+		}
+		if f.Value < 0 {
+			return nil, fmt.Errorf("plotter.NewSankey: Flow %d value (%g) < 0", i, f.Value)
+		}
+
+		// initialize stock holders
+		if _, ok := s.stocks[f.SourceStockCategory]; !ok {
+			s.stocks[f.SourceStockCategory] = make(map[string]*stock)
+		}
+		if _, ok := s.stocks[f.ReceptorStockCategory]; !ok {
+			s.stocks[f.ReceptorStockCategory] = make(map[string]*stock)
+		}
+
+		// figure out plotting order of stocks
+		if _, ok := s.stocks[f.SourceStockCategory][f.SourceStockLabel]; !ok {
+			s.stocks[f.SourceStockCategory][f.SourceStockLabel] = &stock{
+				order:    len(s.stocks[f.SourceStockCategory]),
+				label:    f.SourceStockLabel,
+				category: f.SourceStockCategory,
+			}
+		}
+		if _, ok := s.stocks[f.ReceptorStockCategory][f.ReceptorStockLabel]; !ok {
+			s.stocks[f.ReceptorStockCategory][f.ReceptorStockLabel] = &stock{
+				order:    len(s.stocks[f.ReceptorStockCategory]),
+				label:    f.ReceptorStockLabel,
+				category: f.ReceptorStockCategory,
+			}
+		}
+
+		// add to total value of stocks
+		s.stocks[f.SourceStockCategory][f.SourceStockLabel].sourceValue += f.Value
+		s.stocks[f.ReceptorStockCategory][f.ReceptorStockLabel].receptorValue += f.Value
+	}
+
+	s.LineStyle = DefaultLineStyle
+	s.LineStyle.Color = color.NRGBA{R: 0, G: 0, B: 0, A: 150}
```

Comment:
The lines don't need to be transparent, but the fill does. This can be resolved by taking a parameter for the fill and leaving the decision up to the user.